src/basic/virt.c

```diff
@@ -454,17 +456,117 @@ static const char *const container_table[_VIRTUALIZATION_MAX] = {
 
 DEFINE_PRIVATE_STRING_TABLE_LOOKUP_FROM_STRING(container, int);
 
+static int detect_ancestor_cgroupfs(const char *cgroupfs_mount_point) {
+        int r;
+        _cleanup_fclose_ FILE *proc_self_mountinfo = NULL;
+
+        r = fopen_unlocked("/proc/self/mountinfo", "re", &proc_self_mountinfo);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to open /proc/self/mountinfo: %m");
+
+        for (;;) {
+                _cleanup_free_ char *line = NULL, *root = NULL, *mount_point = NULL;
+                int mount_id, parent_id, major, minor;
+
+                r = read_line(proc_self_mountinfo, LONG_LINE_MAX, &line);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        break;
+
+                if (sscanf(line, "%i %i %i:%i %ms %ms",
+                           &mount_id,
+                           &parent_id,
+                           &major,
+                           &minor,
+                           &root,
+                           &mount_point) != 6)
+                        continue;
+
+                if (!path_equal(mount_point, cgroupfs_mount_point))
+                        continue;
+
+                return !empty_or_root(root);
+        }
+
+        /* We have not found anything mounted at cgroupfs_mount_point. */
+        return false;
+}
+
+static int running_in_cgroupns(void) {
+        int r;
+
+        if (!cg_ns_supported())
+                return false;
+
+        r = cg_all_unified();
+        if (r < 0)
+                return false;
+
+        if (r) {
+                /* cgroup v2 */
+
+                if (detect_ancestor_cgroupfs("/sys/fs/cgroup") > 0)
```

Comment:
Can the order be switched, and do this check last, after the two access() calls? A full parse of mountinfo is going to be expensive when there's loads of mounts. Given this runs on start of every unit with ConditionVirtualization, it's probably going to be worth it.

Reply:
Good idea, although as I understand it `detect_container()` caches its result, so it shouldn't be run *that* much.

https://github.com/systemd/systemd/blob/87cc47ba1d3a4016e84eb1b3f6bddf76266bf978/src/basic/virt.c#L458

Reply:
Ah that's true - but still, better to avoid it at all if possible

Reply:
Moved things around; I'll need to retest if that still works in all the cases.